### ISSUE TITLE
docs(train): document `praisonai-train llm` new `--config`, `--dry-run`, and 11 tuning flags (PraisonAI PR #4362)

### ISSUE THREAD

**Issue by MervinPraison** [MervinPraison/PraisonAIDocs#3349]
## Summary

PraisonAI [PR #4362](https://github.com/MervinPraison/PraisonAI/pull/4362) (merged 2026-08-25 into `main`, head SHA `d54647a`) makes `praisonai-train llm` significantly more usable from the CLI. Before the change, `llm` was the only training subcommand that would **not** accept `--config` (even though `export` and `serve` both do), and it exposed only two flags (`--model`, `--verbose`) out of the ~38 knobs the underlying `unsloth train` accepts — so changing a LoRA rank meant hand-editing a YAML file the CLI gave you no way to point at.

The current docs at `docs/cli/train.mdx` (the `praisonai train llm` section) explicitly say _"Fine-tuning parameters live in the YAML config, not on the CLI"_. **That sentence is now false** and must be replaced.

This is a **documentation UPDATE**, not new content. Two existing files need edits; no new pages are required.

---

## What changed in the SDK (source of truth)

**File:** `src/praisonai-train/praisonai_train/cli/commands/train.py`  
**Test file (spec):** `src/praisonai-train/tests/unit/train/test_llm_cli_flags.py`

### 1. New `--config` / `-c` option

```
--config, -c   Path   Training config YAML. Flags below override it.
```

- `llm` now accepts a `--config path/to/config.yaml` just like `export` and `serve` already did.
- Precedence: **file first, flags on top**. An untouched config key survives when a different key is overridden by a flag (test: `test_a_flag_overrides_the_file`).
- A config file that is **not a YAML mapping** (e.g. a top-level list) is refused up front:  
  `A training config is a YAML mapping of key: value.` — exit `1`.

### 2. New `--dry-run` flag

```
--dry-run   Print the resolved config and exit, without training.
```

- Prints the resolved config as YAML you can paste back into a file, followed by a comment listing which keys came from flags:

  ```
  $ praisonai-train llm -c config.yaml --lora-r 32 --epochs 2 --dry-run
  lora_r: 32
  max_seq_length: 2048
  model_name: unsloth/gemma-2-2b-it-bnb-4bit
  num_train_epochs: 2.0

  # 2 value(s) came from flags: lora_r, num_train_epochs
  ```
- Does **not** load the heavy runner, so `--dry-run` works even without the `[llm]` extras installed.
- `agents` had a `--dry-run` for a while; `llm` (where an hour of rented GPU is at stake) now has one too.

### 3. 11 new tuning flags

Every flag is optional and maps to a `config.yaml` key. A test (`test_every_flag_maps_to_a_config_key_that_exists`) pins the flag → key mapping to `TrainModel.KNOWN_KEYS` so the two cannot drift apart.

| Flag | Config key | Type | Help text |
|------|------------|------|-----------|
| `--method` | `method` | `str` | `sft \| cpt \| dpo \| orpo \| kto` |
| `--max-seq-length` | `max_seq_length` | `int` | Sequence length. |
| `--epochs` | `num_train_epochs` | `float` | Epochs. Ignored if `--max-steps` is set. |
| `--max-steps` | `max_steps` | `int` | Stop after this many steps. |
| `--learning-rate` | `learning_rate` | `float` | Learning rate. |
| `--batch-size` | `per_device_train_batch_size` | `int` | Per-device batch size. |
| `--grad-accum` | `gradient_accumulation_steps` | `int` | Gradient accumulation steps. |
| `--lora-r` | `lora_r` | `int` | LoRA rank. |
| `--lora-alpha` | `lora_alpha` | `int` | LoRA alpha. |
| `--output-dir` | `output_dir` | `str` | Where checkpoints go. |
| `--chat-template` | `chat_template` | `str` | Chat template name. |

Everything else stays in `config.yaml`, which is now reachable via `--config`.

### 4. Dataset argument becomes optional (with a guard)

- The positional `DATASET` argument used to be required. It is now `Optional[str]` so `--config` can supply the dataset.
- If **neither** a positional dataset **nor** a `--config` (with a `dataset:` in it) is given, the CLI refuses with:
  - `error: No dataset given` — `remediation: Pass a dataset path, or --config a file that names one.` — exit `1`.
- Test `test_a_dataset_is_still_required_when_no_config_names_one` pins this — "I forgot the dataset" must not silently train against the default Alpaca corpus.

### 5. Dataset shape normalization (bugfix)

- A `dataset` given as a plain string (from the positional argument or `dataset: ...` in a config) is normalised to `[{"name": "..."}]` before being written to `config.yaml`, because the trainer iterates `config["dataset"]` expecting a mapping per entry. A bare string would otherwise be read character-by-character.
- Test `test_the_dataset_reaches_the_trainer_via_the_written_config` pins this end-to-end.

### 6. Internal: resolved config is materialised to `./config.yaml`

The legacy `train` dispatcher reads only `./config.yaml` in the cwd and parses argv with `parse_known_args()` (silently dropping unknown flags). To make the preview and the real run agree, `_materialize_config()` writes the resolved config to `./config.yaml` before dispatching, and only `train` (plus a belt-and-braces `--dataset ...`) is passed on the command line. Users don't call this directly, but it explains why a `config.yaml` may appear in their cwd after a training run.

---

### 1. `docs/cli/train.mdx` — section `## praisonai train llm`

**Current options table** — only two rows:

| Option | Short | Description | Default |
|--------|-------|-------------|---------|
| `--model` | `-m` | Base model to fine-tune ... | - |
| `--verbose` | `-v` | Verbose output | `false` |

**Current Note (now false):**
> Fine-tuning parameters live in the YAML config, not on the CLI. See [Train → Config.yaml example](/docs/train#configyaml-example) — this now includes `chat_template` ...

**Current signature:**
```bash
praisonai train llm [OPTIONS] DATASET
```
The `DATASET` positional is presented as required.

**Current examples** — no `--config`, no `--dry-run`, no tuning flags.

### 2. `docs/features/praisonai-train-package.mdx` — `## Quick Start` → `Step title="LLM Fine-tuning"`

Every tab (Llama / Gemma / Qwen) uses `praisonai-train llm dataset.json --model ...` only — no demonstration of `--config`, `--dry-run`, or the tuning flags.

---

## Requested doc changes

Both files below live under `docs/` (allowed by AGENTS.md — `docs/cli/` and `docs/features/`).  
**Do not** touch `docs/concepts/`. **Do not** modify auto-generated `docs/js/` or `docs/rust/` pages.

### A. `docs/cli/train.mdx` — `praisonai train llm` section (primary edit)

1. **Rewrite the signature** so `DATASET` is shown as optional and `--config` is the alternative:

   ```bash
   praisonai train llm [OPTIONS] [DATASET]
   ```
   with a one-line rider: _"Pass a dataset path, or `--config` a file that names one — one of the two is required."_

2. **Replace the options table** with the full set (add these rows to the existing `--model` / `--verbose` — keep those; add `--config`, `--dry-run`, and the 11 tuning flags). Match the existing table style used elsewhere in the file.

3. **Delete or rewrite the misleading Note.** The line _"Fine-tuning parameters live in the YAML config, not on the CLI"_ is now wrong. Replace it with a short precedence note:

   > `--config` sets the baseline, the tuning flags override it. Run with `--dry-run` first to see the resolved config the trainer would use — file first, flags on top.

4. **Add a "Precedence" mini-section** (a two-column table or a short prose block) showing the resolution order:

   1. Bare defaults (from `TrainModel`).
   2. `--config config.yaml` (baseline).
   3. Tuning flags on the command line (override).

5. **Add a `--dry-run` example** (verbatim from the PR body reads well):

   ```bash
   praisonai-train llm -c config.yaml --lora-r 32 --epochs 2 --dry-run
   ```

   with the sample output block and a note that `--dry-run` does not load the training runner, so it works even without the `[llm]` extra installed.

6. **Add three new examples** to the Examples list:
   - `praisonai-train llm data.jsonl --lora-r 32 --epochs 2`
   - `praisonai-train llm -c config.yaml --max-seq-length 4096`
   - `praisonai-train llm -c config.yaml --dry-run`

7. **Add a short "Missing dataset" callout** documenting the exit-1 refusal when neither the positional dataset nor a config-supplied one is given.

8. **Cross-link from `--config` to the config-key tables already in the file** (`### Config keys for LLM fine-tuning`) so the reader can see which keys a config file may set beyond the flags.

### B. `docs/features/praisonai-train-package.mdx` — `Quick Start` → `LLM Fine-tuning`

Add a fourth tab (or a short "Or drive it from a config file" block after the model tabs) demonstrating:

```bash
# Preview the resolved config before spending GPU time
praisonai-train llm -c config.yaml --lora-r 32 --epochs 2 --dry-run

# Then run it — flags override the file
praisonai-train llm -c config.yaml --lora-r 32 --epochs 2
```

Keep it agent-centric / beginner-first per AGENTS.md: one sentence framing ("Preview a run before you spend GPU time"), then the two commands.

### C. Mermaid diagram (optional but recommended)

Add a small diagram to `docs/cli/train.mdx` under the `llm` section, showing the precedence, using the standard color scheme from AGENTS.md §3.1:

```mermaid
graph LR
    subgraph "Config resolution"
        Defaults[⚙️ Defaults] --> File[📄 config.yaml]
        File --> Flags[🚩 CLI flags]
        Flags --> Preview[🔍 --dry-run<br/>preview]
        Flags --> Train[🚀 train]
    end

    classDef base fill:#6366F1,stroke:#7C90A0,color:#fff
    classDef file fill:#189AB4,stroke:#7C90A0,color:#fff
    classDef flag fill:#F59E0B,stroke:#7C90A0,color:#fff
    classDef out fill:#10B981,stroke:#7C90A0,color:#fff

    class Defaults base
    class File file
    class Flags flag
    class Preview,Train out
```

---

## Non-goals

- **Do not** create a new page. The changes fit inside the existing `## praisonai train llm` section of `docs/cli/train.mdx` and the existing `Quick Start` of `docs/features/praisonai-train-package.mdx`.
- **Do not** duplicate the full `TrainModel` config reference — the existing `### Config keys for LLM fine-tuning` block already covers that. Only add rows for the 11 flags to the CLI options table.
- **Do not** modify `docs/concepts/` (human-only per AGENTS.md).
- **Do not** touch `docs.json` structure — no new nav entries are needed since we're editing existing pages.

---

## Verification checklist for the follow-up docs PR

- [ ] `docs/cli/train.mdx` options table under `## praisonai train llm` lists all 13 options (`--model`, `--config`/`-c`, `--dry-run`, the 11 tuning flags, `--verbose`).
- [ ] The false Note about "Fine-tuning parameters live in the YAML config, not on the CLI" is removed or rewritten.
- [ ] `[DATASET]` is shown as optional in the signature, and the "missing dataset" refusal is documented.
- [ ] A `--dry-run` example with sample output is included.
- [ ] Precedence (file → flags) is stated at least once, explicitly.
- [ ] `docs/features/praisonai-train-package.mdx` Quick Start LLM step demonstrates at least one `--config` + `--dry-run` invocation.
- [ ] No files under `docs/concepts/`, `docs/js/`, or `docs/rust/` are modified.
- [ ] All new code blocks copy-paste-run against a real `config.yaml` (per AGENTS.md §5.1).

---

## References

- **PraisonAI PR:** [MervinPraison/PraisonAI#4362](https://github.com/MervinPraison/PraisonAI/pull/4362) — `feat(train): give llm a --config, a --dry-run and the flags that matter`
- **Merged commit:** `d54647ae9e123f738b4e7f8eaea6d2ba92b1edc1`
- **Changed files:** `src/praisonai-train/praisonai_train/cli/commands/train.py` (+173 / −12), `src/praisonai-train/tests/unit/train/test_llm_cli_flags.py` (+159, new file).
- **Tests to consult for exact behaviour:** `test_llm_accepts_a_config_file`, `test_a_flag_overrides_the_file`, `test_flags_work_with_no_config_at_all`, `test_a_config_that_is_not_a_mapping_is_refused`, `test_dry_run_does_not_train`, `test_dry_run_says_which_values_came_from_flags`, `test_dry_run_output_is_valid_yaml`, `test_every_flag_maps_to_a_config_key_that_exists`, `test_the_flags_cover_what_a_run_is_usually_tuned_by`, `test_a_dataset_is_still_required_when_no_config_names_one`, `test_the_dataset_reaches_the_trainer_via_the_written_config`.
